File: frontend/src/api/admin/tlsFingerprintProfile.ts
/**
 * Admin TLS Fingerprint Profile API endpoints
 * Handles TLS fingerprint profile CRUD for administrators
 */

import { apiClient REDACTED from '../client'

/**
 * TLS fingerprint profile interface
 */
export interface TLSFingerprintProfile {
  id: number
  name: string
  description: string | null
  enable_grease: boolean
  cipher_suites: number[]
  curves: number[]
  point_formats: number[]
  signature_algorithms: number[]
  alpn_protocols: string[]
  supported_versions: number[]
  key_share_groups: number[]
  psk_modes: number[]
  extensions: number[]
  created_at: string
  updated_at: string
REDACTED

/**
 * Create profile request
 */
export interface CreateProfileRequest {
  name: string
  description?: string | null
  enable_grease?: boolean
  cipher_suites?: number[]
  curves?: number[]
  point_formats?: number[]
  signature_algorithms?: number[]
  alpn_protocols?: string[]
  supported_versions?: number[]
  key_share_groups?: number[]
  psk_modes?: number[]
  extensions?: number[]
REDACTED

/**
 * Update profile request
 */
export interface UpdateProfileRequest {
  name?: string
  description?: string | null
  enable_grease?: boolean
  cipher_suites?: number[]
  curves?: number[]
  point_formats?: number[]
  signature_algorithms?: number[]
  alpn_protocols?: string[]
  supported_versions?: number[]
  key_share_groups?: number[]
  psk_modes?: number[]
  extensions?: number[]
REDACTED

export async function list(): Promise<TLSFingerprintProfile[]> {
  const { data REDACTED = await apiClient.get<TLSFingerprintProfile[]>('/admin/tls-fingerprint-profiles')
  return data
REDACTED

export async function getById(id: number): Promise<TLSFingerprintProfile> {
  const { data REDACTED = await apiClient.get<TLSFingerprintProfile>(`/admin/tls-fingerprint-profiles/${idREDACTED`)
  return data
REDACTED

export async function create(profileData: CreateProfileRequest): Promise<TLSFingerprintProfile> {
  const { data REDACTED = await apiClient.post<TLSFingerprintProfile>('/admin/tls-fingerprint-profiles', profileData)
  return data
REDACTED

export async function update(id: number, updates: UpdateProfileRequest): Promise<TLSFingerprintProfile> {
  const { data REDACTED = await apiClient.put<TLSFingerprintProfile>(`/admin/tls-fingerprint-profiles/${idREDACTED`, updates)
  return data
REDACTED

export async function deleteProfile(id: number): Promise<{ message: string REDACTED> {
  const { data REDACTED = await apiClient.delete<{ message: string REDACTED>(`/admin/tls-fingerprint-profiles/${idREDACTED`)
  return data
REDACTED

export const tlsFingerprintProfileAPI = {
  list,
  getById,
  create,
  update,
  delete: deleteProfile
REDACTED

export default tlsFingerprintProfileAPI

File: backend/internal/domain/tls_fingerprint_profile.go
package domain

import (
	"time"

	"github.com/Wei-Shaw/sub2api/internal/pkg/tlsfingerprint"
)

// TLSFingerprintProfile is a TLS ClientHello fingerprint template.
// Empty slice fields fall back to built-in defaults in the dialer.
type TLSFingerprintProfile struct {
	ID                  int64
	Name                string
	Description         *string
	EnableGREASE        bool
	CipherSuites        []uint16
	Curves              []uint16
	PointFormats        []uint16
	SignatureAlgorithms []uint16
	ALPNProtocols       []string
	SupportedVersions   []uint16
	KeyShareGroups      []uint16
	PSKModes            []uint16
	Extensions          []uint16
	CreatedAt           time.Time
	UpdatedAt           time.Time
}

// Validate reports whether the profile configuration is usable.
func (p *TLSFingerprintProfile) Validate() error {
	if p == nil || p.Name == "" {
		return &ValidationError{Field: "name", Message: "name is required"}
	}
	return nil
}

// ToTLSProfile converts the domain model into a runtime tlsfingerprint.Profile.
// Empty slice fields fall back to built-in defaults in the dialer.
func (p *TLSFingerprintProfile) ToTLSProfile() *tlsfingerprint.Profile {
	if p == nil {
		return nil
	}
	return &tlsfingerprint.Profile{
		Name:                p.Name,
		EnableGREASE:        p.EnableGREASE,
		CipherSuites:        p.CipherSuites,
		Curves:              p.Curves,
		PointFormats:        p.PointFormats,
		SignatureAlgorithms: p.SignatureAlgorithms,
		ALPNProtocols:       p.ALPNProtocols,
		SupportedVersions:   p.SupportedVersions,
		KeyShareGroups:      p.KeyShareGroups,
		PSKModes:            p.PSKModes,
		Extensions:          p.Extensions,
	}
}
